src/unit.c

```diff
@@ -1922,3 +1543,400 @@ SEXP validUnits(SEXP units)
+			}
+			if (Rf_isString(singleData)) {
+				if (!dataCopied) {
+					data = PROTECT(shallow_duplicate(data));
+					dataCopied = 1;
+				}
+				SEXP fcall = PROTECT(lang2(install("gPath"), singleData));
+				singleData = eval(fcall, R_gridEvalEnv);
+				SET_VECTOR_ELT(data, i % nData, singleData);
+				UNPROTECT(1);
+			}
+			if (Rf_inherits(singleData, "gPath")) {
+				SEXP fcall = PROTECT(lang2(install("depth"), singleData));
+				SEXP depth = PROTECT(eval(fcall, R_gridEvalEnv));
+				int tooDeep = INTEGER(depth)[0] > 1;
+				UNPROTECT(2);
+				if (tooDeep) {
+					error(_("'gPath' must have depth 1 in 'grobwidth/height' units"));
+				}
+			}
+		}
+		if (!unitIsString && !unitIsGrob && singleData != R_NilValue) {
+			error(_("non-NULL value supplied for plain unit"));
+		}
+	}
+	UNPROTECT(dataCopied);
+	return data;
+}
+void makeSimpleUnit(SEXP values, SEXP unit) {
+	setAttrib(values, install("unit"), unit);
+	SEXP classes = PROTECT(allocVector(STRSXP, 2));
+	SET_STRING_ELT(classes, 0, mkChar("simpleUnit"));
+	SET_STRING_ELT(classes, 1, mkChar("unit"));
+	classgets(values, classes);
+	UNPROTECT(1);
+}
+SEXP constructUnits(SEXP amount, SEXP data, SEXP unit) {
+	int nAmount = LENGTH(amount);
+	int nData = LENGTH(data);
+	int nUnit = LENGTH(unit);
+	SEXP valUnits = PROTECT(validUnits(unit));
+	if (nUnit == 1) {
+		int u = INTEGER(valUnits)[0];
+		if (!(isStringUnit(u) || isGrobUnit(u))) {
+			int referenced = MAYBE_REFERENCED(amount);
+			if (referenced) amount = PROTECT(duplicate(amount));
+			makeSimpleUnit(amount, valUnits);
+			UNPROTECT(1 + referenced);
+			return amount;
+		}
+	}
+	int n = nAmount < nUnit ? nUnit : nAmount;
+	SEXP units = PROTECT(allocVector(VECSXP, n));
+	
+	data = validData(data, valUnits, n);
+	
+	double* pAmount = REAL(amount);
+	int *pValUnits = INTEGER(valUnits);
+	for (int i = 0; i < n; i++) {
+		SEXP unit = SET_VECTOR_ELT(units, i, allocVector(VECSXP, 3));
+		SET_VECTOR_ELT(unit, 0, Rf_ScalarReal(pAmount[i % nAmount]));
+		SET_VECTOR_ELT(unit, 1, VECTOR_ELT(data, i % nData));
+		SET_VECTOR_ELT(unit, 2, Rf_ScalarInteger(pValUnits[i % nUnit]));
+	}
+	SEXP cl = PROTECT(mkString("unit"));
+	classgets(units, cl);
+	UNPROTECT(3);
+	return units;
+}
+SEXP asUnit(SEXP simpleUnit) {
+	if (inherits(simpleUnit, "unit")) {
+		if (!inherits(simpleUnit, "simpleUnit")) {
+			return simpleUnit;
+		}
+	} else {
+		error(_("object is not coercible to a unit"));
+	}
+	int n = LENGTH(simpleUnit);
+	SEXP units = PROTECT(allocVector(VECSXP, n));
+	double* pAmount = REAL(simpleUnit);
+	SEXP valUnit = getAttrib(simpleUnit, install("unit"));
+	for (int i = 0; i < n; i++) {
+		SEXP unit = SET_VECTOR_ELT(units, i, allocVector(VECSXP, 3));
+		SET_VECTOR_ELT(unit, 0, Rf_ScalarReal(pAmount[i]));
+		SET_VECTOR_ELT(unit, 1, R_NilValue);
+		SET_VECTOR_ELT(unit, 2, valUnit);
+	}
+	SEXP cl = PROTECT(mkString("unit"));
+	classgets(units, cl);
+	UNPROTECT(2);
+	return units;
+}
+SEXP conformingUnits(SEXP unitList) {
```

Comment:
Would you mind giving me a plain-english overview of what this function is doing and why ?  (maybe I'm still tired)

Reply:
Sure - I should probably put it in code as well :-)

It takes a list of units and check if they are all `simpleUnit` object of the same unit type. It is used in e.g. `Ops.summary` and `unit.c` (though `identicalUnits`) to optimise the operations

Reply:
I think I get this now.  Thanks